module/VuFind/src/VuFind/Http/CachingDownloader.php

```diff
@@ -0,0 +1,168 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Http
+ * @author   Mario Trojan <mario.trojan@uni-tuebingen.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+namespace VuFind\Http;
+
+use Laminas\Cache\Storage\StorageInterface;
+use VuFind\Cache\Manager as CacheManager;
+use VuFindHttp\HttpService;
+
+/**
+ * Caching downloader.
+ *
+ * @category VuFind
+ * @package  Http
+ * @author   Mario Trojan <mario.trojan@uni-tuebingen.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class CachingDownloader
+{
+    /**
+     * CacheManager to update caches if necessary.
+     *
+     * @var CacheManager
+     */
+    protected $cacheManager;
+
+    /**
+     * Stored client options for cache key generation.
+     *
+     * @var array
+     */
+    protected $cacheOptions = [];
+
+    /**
+     * Cache to use for downloads
+     *
+     * @var StorageInterface
+     */
+    protected $cache;
+
+    /**
+     * HTTP service
+     *
+     * @var HttpService
+     */
+    protected $httpService;
+
+    /**
+     * Constructor
+     *
+     * @param HttpService $httpService  HTTP service
+     * @param string      $cacheManager Base directory for cache
+     */
+    public function __construct(HttpService $httpService, CacheManager $cacheManager)
+    {
+        $this->httpService = $httpService;
+        $this->cacheManager = $cacheManager;
+        $this->setCacheId('downloader');
+    }
+
+    /**
+     * Set a different cache.
+     *
+     * @param string $cacheId Cache ID
+     *
+     * @return void
+     */
+    public function setCacheId($cacheId)
+    {
+        $cacheName = $this->cacheManager->addDownloaderCache($cacheId);
+        $this->cache = $this->cacheManager->getCache($cacheName);
+    }
+
+    /**
+     * Download a resource using the cache in the background.
+     *
+     * @param string   $url              URL
+     * @param array    $params           Request parameters (e.g. additional headers)
+     * @param callable $validateCallback Callback for validation
+     *                                   before storing to cache
+     * @param callable $decodeCallback   Callback for decoding
+     *
+     * @return mixed
+     */
+    public function download(
+        $url,
+        $params=[],
+        callable $validateCallback=null,
+        callable $decodeCallback=null
+    ) {
+        $cacheItemKey = md5($url);
+        foreach ($params as $paramKey => $paramValue) {
+            $cacheItemKey .= '#' . $paramKey . '§' . $paramValue;
+        }
+
+        // Add new item to cache if not exists
+        if (!$this->cache->hasItem($cacheItemKey)) {
+            $response = $this->httpService->get($url, $params);
```

Comment:
There are actually two aspects here, and they both can be important for meaningful error log messages etc.:

1.) Since the download method never returns a response object, the exception thrown when a request fails should contain all necessary information for the caller to log and handle the error. One option would be a new exception class that would include the url, status code, response headers and body for a failed request, and another for when the HTTP service throws (like when unable to connect) that would include the url and the previous exception.

2.) You need to catch and re-throw to include the url in the exception message, but if the above can alleviate the need.

Reply:
@mtrojan-ub, we can't assume that HttpService will use cURL -- it depends on the adapter setting in the [Http] section of config.ini.

Reply:
@mtrojan-ub if HttpService doesn't use the exception, I don't feel it belongs there.

Reply:
OK, so i added a class `VuFind\Exception\HttpDownloadException` and tried to set additional properties depending on the error scenario. What do you think?